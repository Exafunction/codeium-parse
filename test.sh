#!/bin/bash
set -euo pipefail

cd "$(dirname "${BASH_SOURCE[0]}")"
for test_file in test_files/*; do
    test_file="$(basename "$test_file")"
    echo "$test_file"
    ./parse -file "test_files/$test_file" -use_tags_query -tags_query_dir "queries" > "goldens/$test_file.golden.tmp"
    trap 'rm "goldens/$test_file.golden.tmp"' EXIT
    diff -u "goldens/$test_file.golden" "goldens/$test_file.golden.tmp"
done
