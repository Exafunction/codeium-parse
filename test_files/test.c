#include "runtime/parser.h"

typedef struct {
  Parser* parser;
  TSSymbol lookahead_symbol;
} SkipPrecedingTreesSession;

static void parser__push(Parser* self, StackVersion version, Tree* tree,
                         TSStateId state) {
  ts_stack_push(self->stack, version, tree, false, state);
  ts_tree_release(tree);
}

static Tree* parser__lex(Parser* self, StackVersion version) {
  TSStateId parse_state = ts_stack_top_state(self->stack, version);
  Length start_position = ts_stack_top_position(self->stack, version);
  TSLexMode lex_mode = self->language->lex_modes[parse_state];
}
