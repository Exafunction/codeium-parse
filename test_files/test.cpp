#include <string>

namespace bar {
// Declaration.
bool baz(const std::string& bing);

struct Foo2;

}  // namespace bar

namespace bar::bar2 {
struct Foo3;
}

class Foo {
 public:
  // foo
  // foo2
  void foo() {
    // Do nothing.
  }
};

struct bar::Foo2 {};
struct bar::bar2::Foo3 {};

// baz
bool bar::baz(const std::string& bing) {
  // This shouldn't be recognized as a function.
  Foo dist();
  return bing.empty();
}

// one_pointer comment
bool* one_pointer() { return nullptr; }
// two_pointer comment
bool** two_pointer() { return nullptr; }

// Comment.
template <typename T>
class TemplatedFoo {
  template <typename U>
  void foo();
};

// Comment.
template <typename T>
template <typename U>
void TemplatedFoo<T>::foo() {}

// Comment.
template <>
template <>
void TemplatedFoo<int>::foo<int>() {}

// Comment.
template <typename T>
class TemplatedFoo2;

// Comment.
template <>
class TemplatedFoo2<int> {};
