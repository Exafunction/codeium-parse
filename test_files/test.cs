// clang-format off
namespace Namespace {
  // Comment A.
  internal delegate void A(params int[] test);
  // Comment F.
  public struct F<T>
      where T : struct {
  }
  // Comment F.
  [Nice]
  private record F<T1, T2>
      where T1 : I1, I2, new()
      where T2 : I2 {}
  // Comment Teacher.
  record Teacher(string FirstName, string LastName, string Subject) : Person(FirstName, LastName);
  // Comment B.
  enum B { Ten = 10, Twenty = 20 }
  // Comment F.
  public class F : object, IAlpha, IOmega {}
  // Comment Class.
  public partial class Class<in TParam>
      where TParam : class ?, notnull, F ? {
    // Comment operator +.
    public static int operator +(A a) {
      return 0;
    }
    // Comment GetSet.
    uint GetSet { get; set; }
    // Comment Foo.
    static extern Foo() => bar = 0;
    // Comment ~Class.
    extern ~Class() {}
    // Comment Bar
    public void Bar() => bar = 0;
  }
  // Comment IFoo.
  public interface IFoo {}
}
// clang-format on
