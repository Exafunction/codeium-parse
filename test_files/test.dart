import 'package:test/test.dart';

// Functions

// foo not a docstring
void foo() {
    print("Hello, World!");
}

/// bar is a docstring
int bar() {
    return 0;
}

// Classes with methods

class FooClass {
    int someField = 0;

    // Constructor
    FooClass() {
        this.someField = 0;
    }

    /// FCB docstring
    int fooClassBar() {
        return 0;
    }
}

// Mixins
mixin FooMixin {
    int barMixin() {
        return 0;
    }
}

// Combinations
class FooBar extends FooClass with Foo {
    int bar() {
        return 0;
    }
}

void main() {
    print("Hello, World!");
}
