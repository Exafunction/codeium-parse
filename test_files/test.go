package test_files

// Pattern 0.
type Pattern0 struct {
	a string
}

// Pattern 1.
type Pattern1 interface {
	a() string
}

// Unsupported.
type (
	UnsupportedStruct    struct{}
	UnsupportedInterface interface{}
)

func Pattern2() {}

// Comments
// here.
func (p *Pattern0) Method(string) {}

func (p Pattern0) Method2(a string) string {
	Pattern2()
	s := &Pattern0{}
	s.Method("")
	((*Pattern0)(s)).Method("")
	var s2 any
	s2.(*Pattern0).Method("")
	return a
}

type Alias string
