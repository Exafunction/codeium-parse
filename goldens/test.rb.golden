Name: require
Definition (definition.import):
require "net/http"

Name: SomeClass
Doc:
# This is just your standard class
#  nothing strange here
Declaration (definition.class):
class SomeClass
  # In ruby, the parentheses can be omitted in many circumstances
  # This is probably a job for the AST, but we'll add a test here anyway
  def empty_parens()
    "Nice"
  end
end

Name: empty_parens
Parameters: ()
Definition (definition.method):
def empty_parens()
    "Nice"
  end
Lineage: [SomeClass]
Lineage types: [class]

Name: SomeModule
Definition (definition.module):
module SomeModule
  def omitted_parens
    :foo
  end
end

Name: Test
Definition (definition.module):
module Test ; end

Name: SubTest
Definition (definition.module):
module Test::SubTest
  # We're inheriting from SomeClass here
  class MyClass < SomeClass
    include SomeModule

    # Some languages would call this a "static" method
    def self.in_eigenclass
      new.simple
    end

    class << self
      # In all ways, this method behaves the same as `in_eigenclass`
      def in_reopened_eigenclass
        new.optional_param(true)
      end
    end

    def simple
      self
    end

    def unparenthesized arg1, arg2
      arg2, arg1 = arg1, arg2
    end

    def optional_param(arg1, arg2 = "fine")
      nil
    end

    def splatted_args(arg1, *args)
    end
  end
end

Name: MyClass
Declaration (definition.class):
class MyClass < SomeClass
    include SomeModule

    # Some languages would call this a "static" method
    def self.in_eigenclass
      new.simple
    end

    class << self
      # In all ways, this method behaves the same as `in_eigenclass`
      def in_reopened_eigenclass
        new.optional_param(true)
      end
    end

    def simple
      self
    end

    def unparenthesized arg1, arg2
      arg2, arg1 = arg1, arg2
    end

    def optional_param(arg1, arg2 = "fine")
      nil
    end

    def splatted_args(arg1, *args)
    end
  end
Lineage: [SubTest]
Lineage types: [module]

Name: include
Reference (reference.call):
include SomeModule

Name: simple
Reference (reference.call):
new.simple

Name: optional_param
Reference (reference.call):
new.optional_param(true)

Name: unparenthesized
Parameters: arg1, arg2
Definition (definition.method):
def unparenthesized arg1, arg2
      arg2, arg1 = arg1, arg2
    end
Lineage: [SubTest MyClass]
Lineage types: [module class]

Name: optional_param
Parameters: (arg1, arg2 = "fine")
Definition (definition.method):
def optional_param(arg1, arg2 = "fine")
      nil
    end
Lineage: [SubTest MyClass]
Lineage types: [module class]

Name: SubSubTest
Definition (definition.module):
module Test::SubTest::SubSubTest
  module SubSubSubTest
    extend SomeModule
  end

end

Name: SubSubSubTest
Definition (definition.module):
module SubSubSubTest
    extend SomeModule
  end
Lineage: [SubSubTest]
Lineage types: [module]

Name: extend
Reference (reference.call):
extend SomeModule

Name: MyTerribleMethod
Reference (reference.call):
MyTerribleMethod()

Name: new
Reference (reference.call):
Test::SubTest::MyClass.new

Name: splatted_args
Reference (reference.call):
Test::SubTest::MyClass.new.splatted_args 1, 2, 3

Name: in_reopened_eigenclass
Reference (reference.call):
Test::SubTest::MyClass.in_reopened_eigenclass

Name: omitted_parens
Reference (reference.call):
Test::SubTest::SubSubTest::SubSubSubTest.omitted_parens
